File: JavaScript/java_script/src/venda/venda.service.ts
import { Injectable } from '@nestjs/common';
import { MysqlService } from '../mysql.service';
import { CreateVendaDto } from '../create_.dto';

@Injectable()
export class VendasService {
  constructor(private readonly mysqlService: MysqlService) {}

  async create(dto: CreateVendaDto): Promise<any> {
    const { produto_id, quantidade, preco_unitario, data_venda } = dto;

    const sql = `
      INSERT INTO vendas (
        produto_id, quantidade, preco_unitario, data_venda
      ) VALUES (?, ?, ?, ?)
    `;

    const params = [produto_id, quantidade, preco_unitario, data_venda];
    await this.mysqlService.query(sql, params);

    return { message: 'Venda registrada com sucesso' };
  }

  async findAll(): Promise<any[]> {
    const sql = `
      SELECT v.*, p.nome AS nome_produto
      FROM vendas v
      JOIN produtos p ON p.id = v.produto_id
      ORDER BY v.data_venda DESC
    `;
    return this.mysqlService.query(sql);
  }
  async gerarRelatorioVendasMensais(): Promise<any[]> {
    // 1. Buscar todos os dados de vendas e produtos
    const sql = `
      SELECT
        v.data_venda,
        v.quantidade,
        v.preco_unitario,
        p.categoria
      FROM vendas v
      JOIN produtos p ON p.id = v.produto_id
    `;

    const vendas = await this.mysqlService.query(sql);

    // 2. Processar os dados no código
    const agrupado = new Map<
      string,
      {
        ano: number;
        mes: string;
        categoria: string;
        quantidade_total: number;
        valor_total: number;
      }
    >();

    for (const venda of vendas) {
      const data = new Date(venda.data_venda);
      const ano = data.getFullYear();
      const mes = String(data.getMonth() + 1).padStart(2, '0');
      const categoria = venda.categoria;
      const chave = `${ano}-${mes}-${categoria}`;
      const valor = venda.quantidade * venda.preco_unitario;

      const acumulado = agrupado.get(chave) ?? {
        ano,
        mes,
        categoria,
        quantidade_total: 0,
        valor_total: 0,
      };

      acumulado.quantidade_total += venda.quantidade;
      acumulado.valor_total += valor;

      agrupado.set(chave, acumulado);
    }

    // 3. Gerar projeções e preparar para inserção
    const taxa = 0.015;
    const meses = 12;

    const relatorioFinal = Array.from(agrupado.values()).map((grupo) => ({
      ...grupo,
      projecao_12_meses: parseFloat(
        (grupo.valor_total * Math.pow(1 + taxa, meses)).toFixed(2),
      ),
    }));

    // 4. Escrever no banco (inserir linha por linha)
    for (const item of relatorioFinal) {
      await this.mysqlService.query(
        `
        INSERT INTO relatorio_vendas_mensais (
          ano, mes, categoria, quantidade_total,
          valor_total, projecao_12_meses
        ) VALUES (?, ?, ?, ?, ?, ?)
      `,
        [
          item.ano,
          item.mes,
          item.categoria,
          item.quantidade_total,
          item.valor_total,
          item.projecao_12_meses,
        ],
      );
    }

    return relatorioFinal;
  }
}

File: Java/src/main/java/com/example/api/venda/VendaService.java
package com.example.api.venda;

import com.example.api.Database;

import com.example.api.dto.CreateVendaDto;
import org.springframework.stereotype.Service;

import java.sql.*;
import java.util.*;
import java.util.Date;

@Service
public class VendaService {

    public Map<String, Object> create(CreateVendaDto dto) {
        String sql = """
                    INSERT INTO vendas (
                      produto_id, quantidade, preco_unitario, data_venda
                    ) VALUES (?, ?, ?, ?)
                """;

        try (Connection conn = Database.getConnection();
                PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setInt(1, dto.produto_id);
            stmt.setInt(2, dto.quantidade);
            stmt.setDouble(3, dto.preco_unitario);
            stmt.setString(4, dto.data_venda);
            stmt.executeUpdate();

            Map<String, Object> res = new HashMap<>();
            res.put("message", "Venda registrada com sucesso");
            return res;

        } catch (SQLException e) {
            throw new RuntimeException("Erro ao registrar venda", e);
        }
    }

    public List<Map<String, Object>> findAll() {
        List<Map<String, Object>> vendas = new ArrayList<>();
        String sql = """
                    SELECT v.*, p.nome AS nome_produto
                    FROM vendas v
                    JOIN produtos p ON p.id = v.produto_id
                    ORDER BY v.data_venda DESC
                """;

        try (Connection conn = Database.getConnection();
                PreparedStatement stmt = conn.prepareStatement(sql);
                ResultSet rs = stmt.executeQuery()) {

            while (rs.next()) {
                Map<String, Object> venda = new HashMap<>();
                venda.put("id", rs.getInt("id"));
                venda.put("produto_id", rs.getInt("produto_id"));
                venda.put("quantidade", rs.getInt("quantidade"));
                venda.put("preco_unitario", rs.getDouble("preco_unitario"));
                venda.put("data_venda", rs.getString("data_venda"));
                venda.put("nome_produto", rs.getString("nome_produto"));
                vendas.add(venda);
            }

        } catch (SQLException e) {
            throw new RuntimeException("Erro ao buscar vendas", e);
        }

        return vendas;
    }

    public List<Map<String, Object>> gerarRelatorioVendasMensais() {
        String sql = """
                    SELECT
                      v.data_venda,
                      v.quantidade,
                      v.preco_unitario,
                      p.categoria
                    FROM vendas v
                    JOIN produtos p ON p.id = v.produto_id
                """;

        try (Connection conn = Database.getConnection();
                PreparedStatement stmt = conn.prepareStatement(sql);
                ResultSet rs = stmt.executeQuery()) {

            Map<String, Map<String, Object>> agrupado = new HashMap<>();

            while (rs.next()) {
                Date dataVenda = rs.getDate("data_venda");
                Calendar calendar = Calendar.getInstance();
                calendar.setTime(dataVenda);
                int ano = calendar.get(Calendar.YEAR);
                String mes = String.format("%02d", calendar.get(Calendar.MONTH) + 1);
                String categoria = rs.getString("categoria");
                String chave = ano + "-" + mes + "-" + categoria;

                double valor = rs.getInt("quantidade") * rs.getDouble("preco_unitario");

                Map<String, Object> grupo = agrupado.getOrDefault(chave, new HashMap<>());
                grupo.put("ano", ano);
                grupo.put("mes", mes);
                grupo.put("categoria", categoria);
                grupo.put("quantidade_total",
                        (int) grupo.getOrDefault("quantidade_total", 0) + rs.getInt("quantidade"));
                grupo.put("valor_total", (double) grupo.getOrDefault("valor_total", 0.0) + valor);

                agrupado.put(chave, grupo);
            }

            double taxa = 0.015;
            int meses = 12;

            List<Map<String, Object>> relatorioFinal = new ArrayList<>();

            for (Map<String, Object> grupo : agrupado.values()) {
                double valorTotal = (double) grupo.get("valor_total");
                double projecao = valorTotal * Math.pow(1 + taxa, meses);
                grupo.put("projecao_12_meses", Math.round(projecao * 100.0) / 100.0);
                relatorioFinal.add(grupo);
            }

            // Inserção no banco (opcional)
            for (Map<String, Object> item : relatorioFinal) {
                try (PreparedStatement insertStmt = conn.prepareStatement("""
                            INSERT INTO relatorio_vendas_mensais (
                              ano, mes, categoria, quantidade_total,
                              valor_total, projecao_12_meses
                            ) VALUES (?, ?, ?, ?, ?, ?)
                        """)) {
                    insertStmt.setInt(1, (int) item.get("ano"));
                    insertStmt.setString(2, (String) item.get("mes"));
                    insertStmt.setString(3, (String) item.get("categoria"));
                    insertStmt.setInt(4, (int) item.get("quantidade_total"));
                    insertStmt.setDouble(5, (double) item.get("valor_total"));
                    insertStmt.setDouble(6, (double) item.get("projecao_12_meses"));
                    insertStmt.executeUpdate();
                }
            }

            return relatorioFinal;

        } catch (SQLException e) {
            throw new RuntimeException("Erro ao gerar relatório", e);
        }
    }
}
